### ISSUE TITLE
Following layer 2 tutorial, don't get IP

### ISSUE THREAD

**Issue by carlpett** [danderson/metallb#281]
**Is this a bug report or a feature request?**:
Bug report

**What happened**:
I followed the layer 2 tutorial, after creating nginx service, External IP stays in pending.

**What you expected to happen**:
That I would get an IP

**How to reproduce it (as minimally and precisely as possible)**:
Follow tutorial

**Anything else we need to know?**:
Speaker logs: https://gist.github.com/carlpett/c9bcf8761e5bb93931f102085ef167ba
There is an error, but seems related to IPv6 only?

Completely fresh cluster deployed with kubeadm and Flannel (v0.10.0), two nodes, one master and one agent. VMs running on a single-node Vmware ESXi.

_(Probably unrelated, but for completeness, the CoreDNS deployment is crashlooping due to being unable to reach the kube-api with no route to host - I don't **think** that happened before I deployed metallb, but I'm not sure, also can't really see why it would be impacted. Most likely something with flannel.)_

**Environment**:
- MetalLB version: 0.7.0
- Kubernetes version: 1.11.1
- BGP router type/version: -
- OS (e.g. from /etc/os-release): CentOS Linux 7 (Core)
- Kernel (e.g. `uname -a`): Linux <host> 3.10.0-862.el7.x86_64 #1 SMP Fri Apr 20 16:44:24 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by danderson** [comment]
Thanks for the report. I don't see anything in the logs, which indicates that MetalLB is deciding to not announce the nginx service for some reason. What does `kubectl describe svc nginx` say?

**Comment by carlpett** [comment]
Sorry for missing that in the report. It doesn't say anything. `kubectl get events` also doesn't show anything related to metallb.

**Comment by carlpett** [comment]
Hm, now it works. The IP was assigned when I disabled the firewall to see if that would get CoreDNS working. So, something in the pod networking wasn't happy, which, I guess, caused MetalLB to be unable to assing the IP?
Shouldn't this turn up in the logs, though? Both the controller and speaker are running on the same node, so also unsure what flannel could have impacted then.

**Comment by asvillamar** [comment]
I got it to work by disabling firewalld

**Comment by ghaering** [comment]
Well it seems this issue can be closed, then?

**Comment by carlpett** [comment]
It would seem to me that the diagnosability part of this still warrants keeping this issue open? If MetalLB cannot communicate and thus not work, I think there should be some form of warning? And hopefully even some indication what the user needs to do to fix it?

**Comment by danderson** [comment]
I agree that troubleshooting is currently bad. The problem is that MetalLB cannot know that it's being blocked by a firewall, either on-host or somewhere else on the network.

However, I have plans to at least write some troubleshooting documentation to help people figure out "is MetalLB broken, or is it my network?", and possibly also identify the exact cause (firewalls, asymetric routing, reverse path filtering, ...).

In this case, my guess is the problem was that the controller (not the speaker) could not communicate with the k8s apiserver at all, so it just wasn't doing anything. There's no checking for that right now, and I'm not even sure there's any mechanism for heartbeats/watchdogs in the k8s client, but I should definitely add some logic to detect such communication failures.

Do those two action items sound good to you? If so, I'll file those specific bugs (and/or retitle this one).

**Comment by carlpett** [comment]
On a high level, yep, that sounds good! Agree that it's probably hard to diagnose all the possible things that could be wrong in the network (quite a lot of those...).
Would it be reasonable to have some sort of pinging from the speakers to the controller(s)? If they could log when that doesn't work, and the controller keeps track if it can talk to the apiserver, then you'd at least know if something doesn't work, even though it can't tell you directly why?

**Comment by 13567436138** [comment]
I also met this problem.and my firewalld is diabled.
'''
[root@node1 ~]# kubectl get cm config -n lb -o yaml
apiVersion: v1
data:
  config: |
    address-pools:
    - name: my-ip-space
      protocol: layer2
      addresses:
      - 172.20.0.0/16
kind: ConfigMap
metadata:
  creationTimestamp: 2019-02-01T01:12:35Z
  name: config
  namespace: lb
  resourceVersion: "1162494"
  selfLink: /api/v1/namespaces/lb/configmaps/config
  uid: 75523320-25be-11e9-b3a9-00163e001b45
'''

**Comment by 13567436138** [comment]
is my config wrong

**Comment by 13567436138** [comment]
{"caller":"reflector.go:341","level":"warn","msg":"go.universe.tf/metallb/internal/k8s/k8s.go:237: watch of *v1.ConfigMap ended with: too old resource version: 1599012 (1599627)","ts":"2019-02-03T06:17:06.585589Z"}

**Comment by donkeysharp** [comment]
Hello I have the same issue, 

I am using Centos7 for master and nodes in the cluster. The problem is that it does not assign an IP address at first, but after I disable `firewalld` on the nodes it works and assign an IP.

I am using layer2 settings. Not really sure how `firewalld` works on detail, but I assume it blocks outgoing gratuitous packets that MetalLB sends, I will keep investigating and let you know if I find something.

**Comment by stobias123** [comment]
Same issue - centos7 metallb version `0.7.3` - `firewalld` off

**Comment by scoonie7** [comment]
Same issue.
speaker log seems to be fine but the controller log says:

```
{"caller":"reflector.go:205","level":"error","msg":"go.universe.tf/metallb/internal/k8s/k8s.go:231: Failed to list *v1.Service: Get https://10.96.0.1:443/api/v1/services?limit=500\u0026resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout","ts":"2019-04-23T14:57:48.491485Z"}

{"caller":"reflector.go:205","level":"error","msg":"go.universe.tf/metallb/internal/k8s/k8s.go:237: Failed to list *v1.ConfigMap: Get https://10.96.0.1:443/api/v1/namespaces/metallb-system/configmaps?fieldSelector=metadata.name%3Dconfig\u0026limit=500\u0026resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout","ts":"2019-04-23T14:58:19.273495Z"}
```

**Comment by danderson** [comment]
Sorry, but I no longer have the time to debug MetalLB deployments for people. I want to keep this tracker specifically for bugs and features, so I'm going to close this. At some point I may have time to write general troubleshooting documentation, but as a first step towards that I need to clear up the issue tracker.

You could try asking for help on the #metallb channel on the Kubernetes slack, where there are people other than me who can assist.

**Comment by cedricve** [comment]
having same issue, trying to use this on an ARM device (RPi). Anyone an idea?

**Comment by imranrazakhan** [comment]
@ghost I am facing same issue, Are you able to fix it?
metallb version = 0.8.3

`E0311 12:19:40.290974       1 reflector.go:125] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: Failed to list *v1.ConfigMap: Get https://10.96.0.1:443/api/v1/namespaces/metallb-system/configmaps?fieldSelector=metadata.name%3Dconfig&limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: no route to host`

**Comment by MrAmbiG** [comment]
> @ghost I am facing same issue, Are you able to fix it?
> metallb version = 0.8.3
> 
> `E0311 12:19:40.290974 1 reflector.go:125] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: Failed to list *v1.ConfigMap: Get https://10.96.0.1:443/api/v1/namespaces/metallb-system/configmaps?fieldSelector=metadata.name%3Dconfig&limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: no route to host`

yes, i have posted my answer there on top

**Comment by Aeris126** [comment]
> Please go to cluster settings>advanced>ingress>false.
> Now recreate services and metallb will auto assign ip addresses.
> Prior to this my LB services were always in pending status and no external IP was shown. Now it is all fine. Basically you cant use the rancher's built in nginx ingress and metal lb. Both are a conflict. You disable the build in ingress and metal lb will take over.

Are` you talking about some Metallb web page? or this is nginx-ingress web page? or kubernetes dashboard? please clarify

**Comment by danicheeta** [comment]
I have same problem but disabling firewalld doesn't help.

**Comment by fedepaol** [comment]
@danicheeta mind sharing details about your cluster / logs of the speakers / controller?
Also, it would be good if you can file another issue with such information.
